### ISSUE TITLE
Need tests to ensure 2-way index/codec compatibility with Lucene

### ISSUE THREAD

**Issue by clambertus** [apache/lucenenet#270]
<p>Lucene came with several tests to ensure backward compatibility with indexes, and included many zipped index archives which are used to ensure Lucene.NET can read indexes that were produced by Lucene.</p>

<p>However, we have no tests to ensure that indexes produced by Lucene.NET 4.8.0 can be read by Lucene 4.8.0.</p>

<p>A way this could be done:</p>

<p>1. Create a command line utility that is part of the Lucene.NET build that produces a series of test index cases (at least 1 case per codec/doc values format/postings format combination). Unlike the existing compatibility tests that zip the indexes into embedded resources, we should be creating indexes based on the current build of Lucene.NET.<br/>
2. Create a Java/JUnit test project that depends on Lucene 4.8.0 and add a test per test case.<br/>
3. Add build/test tasks to <tt>azure-pipelines.yml</tt> to run the .NET utility to produce the indexes and then execute Java/JUnit tests on each target framework/OS/platform</p>

<p>These tests should be designed in such a way that when we upgrade to the next version of Lucene the tests are simple to upgrade as well.</p>
<i>JIRA link - <a href="https://issues.apache.org/jira/browse/LUCENENET-613">[LUCENENET-613]</a> created by nightowl888</i>

**Comment by rclabo** [comment]
I had a similar thought and Shad pointed me to this issue.  I was thinking about it slightly differently.  Here is another way we might accomplish the same goal of ensuring that a Lucene.NET 4.8 index is compatible with Lucene 4.8.0:

We could index a decent size set of documents using some test code that utilizes stored fields, DocValues, term positions, norms and such.  Then port that code to java and index the same set of documents using 4.8 java Lucene.  Then run a byte by byte comparison of the index files of Lucene.NET and Java Lucene.  This would unequivocally validate that the codec is sound.  And the cool thing is the code could be leveraged again any time we port another codec like the one used in 9.0.

**Comment by rclabo** [comment]
It was brought to my attention that some of the files may have headers that store such things as "javaVendor" and what we put there for Lucene.NET 4.8 isn't going to match what Lucene 4.8 puts there.  So it seems Java Lucene  intended two files created by two different java vendors would have different meta data., So maybe byte by byte comparison of the headers isn't the right way to go.  Maybe for the hearders of such a file we have to do more sophisticated checking but the rest of the file and other files can be byte by byte comparisons, perhaps.  

I have opened a Lucene.NET 4.8 index in Luke and it didn't throw but I didn't validate every aspect of the file and there are many. A byte by byte comparison of the parts that can be compared would give us a ton of confidence.  Or if the file has a CRC that doesn't include the headers comparing that may be a great way to go.

**Comment by eladmarg** [comment]
why do we care? seriously, no one is going to use same java index in lucene.net

as long as the functionality is working, I think we should put our efforts in other areas such as performance and stability. 
its okay that the lucene.net will not be compatible with the java variant, I'm totally fine by this.
besides, indexes are based on persisted other data, so they can always be rebuild/recreated.

**Comment by rclabo** [comment]
@eladmarg Thanks, I appreciate your thoughts.

I think there are several reasons we'd care that Lucene.NET 4.8 outputs an index that is identical to Java Lucene 4.8 (except for obvious headers) and stability is certainly one of them.  Top of my mind I think we should care because:

1) It absolutely guarantees we don't have a bug in the ported codecs which leads to a solid more stable Lucene.NET
2) It means that users of Lucene.NET can have confidence using other tools designed to work with Lucene such as Luke.
3) If the test is ran on the various platforms that Lucene.NET supports it demonstrates that there aren't portability issue with moving a file between the platforms that Lucene.NET supports.  (Full framework, .NET Core Windows, .NET Core Unix.)

**Comment by NightOwl888** [comment]
Agreed. Although, since Lucene already has created a CheckIndex utility to confirm compatibility using hashes, we should just leverage that rather than trying to mock a way to make our index byte-by-byte compatible with one from Java. The gap that could potentially happen is small - we are already using zipped index files that were created in Java to test with, so we know that binary reading is compatible. However, there is a small chance that we have introduced the same binary writing bug in both the production code and the tests and it would be good to confirm that the CheckIndex util in Java works to load indexes that were created in .NET on a continual basis.

The idea is to build a test module once for this, and use it to confirm binary compatibility with Lucene in each release after 4.8.0. Provided Lucene doesn't radically change their design again, this will be the last full port of Lucene to .NET - all future Lucene.NET releases will just be an upgrade of this one. We will be able to keep pace with Lucene's release schedule, but that will only happen if we automate time-consuming manual processes like checking for binary compatibility and make them part of our nightly build.

Since breaking binary compatibility in a patch is unacceptable, this absolutely needs to be done prior to the 4.8.0 release. We don't really think that any new Java applications will load indexes from Lucene.NET 4.8.0, but as @rclabo pointed out, there are index utilities that exist in the wild that we need to be compatible with.
